Implement account permission delegation - XLS-74d and XLS-75d amendments

Patel-Raj11 opened the pull request:
## High Level Overview of Change

This PR is continuation of https://github.com/XRPLF/xrpl-py/pull/833 as there are more changes related to binary codec involved.

This PR implements changes pertaining to the [XLS-74d and XLS-75d amendments](https://github.com/XRPLF/XRPL-Standards/pull/257/files).

Here is the associated cpp PR:  https://github.com/XRPLF/rippled/pull/5354.

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Did you update CHANGELOG.md?

- [x] Yes
- [ ] No, this change does not impact library users

## Test Plan

Added unit and integration tests.

mvadari commented on xrpl/models/transactions/delegate_set.py:
```diff
@@ -0,0 +1,179 @@
+PERMISSION_MAX_LENGTH = 10
+
+
+class DelegatableTransaction(str, Enum):
```
This should use the `Transaction` class as a base to make maintenance easier

mvadari replied:
Not everything has to be checked at the typing level IMO. This will be much more difficult to maintain, and require a separate set of imports.

Patel-Raj11 replied:
@mvadari @khancode I have addressed this in [3626419](https://github.com/XRPLF/xrpl-py/pull/840/commits/36264195b72232d428ad7233e5d91cb88d582be0). Can you please review?

Patel-Raj11 commented on xrpl/core/binarycodec/types/st_object.py:
```diff
@@ -79,6 +99,12 @@ def _enum_to_str(field: str, value: int) -> Union[str, int]:
+        if value in _GRANULAR_PERMISSIONS.values():
+            for k, v in _GRANULAR_PERMISSIONS.items():
+                if v == value:
+                    return k
```
@mvadari Addressed it in [f2a496d](https://github.com/XRPLF/xrpl-py/pull/840/commits/f2a496d628af9c5e76f9af05524402803fb9ab50)

Patel-Raj11 commented on xrpl/core/binarycodec/definitions/definitions.json:
```diff
@@ -2170,6 +2190,16 @@
       }
     ],
+    [
+      "Permission",
+      {
+        "isSerialized": true,
+        "isSigningField": true,
+        "isVLEncoded": false,
+        "nth": 15,
+        "type": "STObject"
+      }
+    ],
```
Serialization of `Permission` object works as expected as tested by submitting transactions using integration test cases.

Patel-Raj11 commented on xrpl/models/transactions/delegate_set.py:
```diff
@@ -0,0 +1,179 @@
+        return None
+
+    def _get_permissions_error(self: Self) -> Optional[str]:
+        if len(self.permissions) > PERMISSION_MAX_LENGTH:
```
@khancode Empty permissions is required as per the spec to revoke all the permissions. Since there is no DelegateUnset transaction type.

khancode replied:
Ah I see, looks good then :)

Patel-Raj11 commented on tests/integration/transactions/test_delegate_set.py:
```diff
@@ -0,0 +1,201 @@
+        self.assertEqual(response.result["tx_json"]["Delegate"], bob.address)
+        self.assertEqual(response.result["tx_json"]["SigningPubKey"], bob.public_key)
+
+        # Use the bob's account to execute a transaction on behalf of alice
+        account_set = AccountSet(
+            account=alice.address, delegate=bob.address, email_hash=EMAIL_HASH
+        )
+        response = await sign_and_reliable_submission_async(
+            account_set, bob, client, check_fee=False
+        )
+        self.assertEqual(response.status, ResponseStatus.SUCCESS)
+        self.assertEqual(response.result["engine_result"], "tecNO_PERMISSION")
```
Since, we delegated `AccountDomainSet` but we are trying to set `email_hash`

khancode replied:
Oh I see. It would have to be `AccountEmailHashSet` instead.

mvadari commented on xrpl/models/transactions/delegate_set.py:
```diff
@@ -0,0 +1,179 @@
+    XCHAIN_MODIFY_BRIDGE = "XChainModifyBridge"
+
+
+class GranularPermission(str, Enum):
```
The whole point of serialization is to convert strings to numbers. I don't think that needs to be called out specifically.

mvadari commented on xrpl/core/binarycodec/definitions/definitions.py:
```diff
@@ -258,3 +276,35 @@ def get_ledger_entry_type_name(ledger_entry_type: int) -> str:
+    Returns:
+        An integer representing the given permission value string.
+    """
+    if permission_value in _GRANULAR_PERMISSIONS:
+        return _GRANULAR_PERMISSIONS[permission_value]
+
+    return get_transaction_type_code(permission_value) + 1
```
Couldn't this be simplified if instead of having a `_GRANULAR_PERMISSIONS` map, the map included the transaction types?

Patel-Raj11 replied:
@mvadari I came up with this [c36fa12](https://github.com/XRPLF/xrpl-py/pull/840/commits/c36fa12aec7aabdc54e834925147f1467ee54db5). Let me know if there is a more cleaner way to do this.

khancode replied:
Wouldn't that cause confusion since they're not actual transaction type names?

Patel-Raj11 replied:
The map is being referred as `_DELEGATABLE_PERMISSIONS_STR_TO_CODE_MAP` and `_DELEGATABLE_PERMISSIONS_CODE_TO_STR_MAP`. It encompasses both granular and transaction type permissions. I think its clear enough.

khancode replied:
I was responding to @mvadari comment :)

mvadari replied:
> Wouldn't that cause confusion since they're not actual transaction type names?

What do you mean? The transaction type permissions are transaction types?

I'm just suggesting doing something like 
```
_PERMISSIONS = {
  **_TRANSACTION_TYPES,
  "AccountDomainSet",
  ...
}
```

Instead of needing separate lists for the transaction types and permissions and needing to do two lookups.

Patel-Raj11 replied:
@mvadari it's addressed in commit https://github.com/XRPLF/xrpl-py/pull/840/commits/c36fa12aec7aabdc54e834925147f1467ee54db5. Can you please resolve this conversation if its good.

khancode replied:
@mvadari Ah I see, I thought you meant to add the granular permissions inside `TRANSACTION_TYPES` too. Looks good to have all included under a permissions variable.

Patel-Raj11 commented on xrpl/models/transactions/delegate_set.py:
```diff
@@ -0,0 +1,140 @@
+
+PERMISSIONS_MAX_LENGTH = 10
+
+NON_DELEGATABLE_TRANSACTIONS = {
+    TransactionType.ACCOUNT_SET.value,
+    TransactionType.SET_REGULAR_KEY.value,
+    TransactionType.SIGNER_LIST_SET.value,
+    TransactionType.DELEGATE_SET.value,
+    TransactionType.ACCOUNT_DELETE.value,
+}
```
IMO it's better to have this check so we don't waste a network requests for client library users. Since, we have checks on max lengths to prevent sending requests that we know will fail, it makes sense to keep it.

khancode replied:
That's fair 👍

Patel-Raj11 commented on tests/integration/transactions/test_delegate_set.py:
```diff
@@ -0,0 +1,201 @@
+        self.assertEqual(response.status, ResponseStatus.SUCCESS)
+        self.assertEqual(response.result["engine_result"], "tesSUCCESS")
+
+        ledger_entry_response = await client.request(
+            LedgerEntry(
+                delegate=Delegate(
+                    account=alice.address,
+                    authorize=bob.address,
+                ),
+            )
+        )
+        self.assertTrue(ledger_entry_response.is_successful())
+        self.assertEqual(
+            ledger_entry_response.result["node"]["LedgerEntryType"],
+            "Delegate",
+        )
+        self.assertEqual(ledger_entry_response.result["node"]["Account"], alice.address)
+        self.assertEqual(ledger_entry_response.result["node"]["Authorize"], bob.address)
+        self.assertEqual(len(ledger_entry_response.result["node"]["Permissions"]), 2)
```
Addressed in [c536ef5](https://github.com/XRPLF/xrpl-py/pull/840/commits/c536ef59fb77366bb19881603bd8173c40fe173c)

khancode commented on xrpl/models/transactions/delegate_set.py:
```diff
@@ -0,0 +1,140 @@
+
+        if set(entered_permissions) & NON_DELEGATABLE_TRANSACTIONS:
+            return (
+                f"Non delegatable transactions found in `permissions` list: "
```
```suggestion
                f"Non-delegatable transactions found in `permissions` list: "
```
nit

Patel-Raj11 commented on xrpl/core/binarycodec/definitions/definitions.json:
```diff
@@ -2560,6 +2590,16 @@
       }
     ],
+    [
+      "Permissions",
+      {
+        "isSerialized": true,
+        "isSigningField": true,
+        "isVLEncoded": false,
+        "nth": 29,
+        "type": "STArray"
+      }
+    ],
```
Such constraints are not added in `definitions.json` file for any of the objects. `definitions.json` is a generated file.

khancode commented on xrpl/models/transactions/delegate_set.py:
```diff
@@ -0,0 +1,140 @@
+    transaction.
+    """
+
+    permission_value: str = REQUIRED  # type: ignore
```
Shouldn't this include `GranularPermission` type that's defined above it?

Patel-Raj11 replied:
We can pass in `TransactionType` and `GranularPermission` as done in unit test cases, and since we didn't want type checks its kept as `str`

Patel-Raj11 replied:
For reference, the types were removed to address this conversation https://github.com/XRPLF/xrpl-py/pull/840#discussion_r2080474360

Patel-Raj11 replied:
My bad, I got what you mean by including `GranularPermission` type. I am making a change to add `GranularPermission` and `TransactionType` so that client library users get IDE suggestions to use these enums available.

Patel-Raj11 replied:
Addressed in [0a85fa1](https://github.com/XRPLF/xrpl-py/pull/840/commits/0a85fa12413ee285dce6db59b9fef93721103c64)

mvadari commented on tests/unit/core/binarycodec/types/test_serialized_dict.py:
```diff
@@ -1,54 +1,37 @@
     def test_from_value(self):
-        transaction = STObject.from_value(expected_json)
-        self.assertEqual(buffer, str(transaction).upper())
+        for test in _FIXTURES_JSON:
```
I recommend using `self.subTest` for these, for better error messaging

mvadari commented on xrpl/core/binarycodec/definitions/definitions.py:
```diff
@@ -59,6 +59,29 @@ def load_definitions(filename: str = "definitions.json") -> Dict[str, Any]:
+    "MPTokenIssuanceLock": 65547,
+    "MPTokenIssuanceUnlock": 65548,
+}
+_DELEGATABLE_PERMISSIONS_STR_TO_CODE_MAP: Dict[str, int] = {
+    **{key: value + 1 for (key, value) in _DEFINITIONS["TRANSACTION_TYPES"].items()},
+    **{key: value for (key, value) in _GRANULAR_PERMISSIONS.items()},
+}
+_DELEGATABLE_PERMISSIONS_CODE_TO_STR_MAP: Dict[int, str] = {
+    **{value + 1: key for (key, value) in _DEFINITIONS["TRANSACTION_TYPES"].items()},
+    **{value: key for (key, value) in _GRANULAR_PERMISSIONS.items()},
+}
```
```suggestion
_DELEGATABLE_PERMISSIONS_STR_TO_CODE_MAP: Dict[str, int] = {
    **{key: value + 1 for (key, value) in _DEFINITIONS["TRANSACTION_TYPES"].items()},
    **_GRANULAR_PERMISSIONS,
}
_DELEGATABLE_PERMISSIONS_CODE_TO_STR_MAP: Dict[int, str] = {
    **{value + 1: key for (key, value) in _DEFINITIONS["TRANSACTION_TYPES"].items()},
    **_GRANULAR_PERMISSIONS,
}
```

mvadari commented on xrpl/models/transactions/delegate_set.py:
```diff
@@ -0,0 +1,149 @@
+    transaction.
+    """
+
+    permission_value: Union[
+        TransactionType,
+        GranularPermission,
+        str,
+    ] = REQUIRED  # type: ignore
```
```suggestion
    permission_value: Union[
        TransactionType,
        GranularPermission
    ] = REQUIRED  # type: ignore
```

This would also simplify the logic below - you'd no longer need `_normalize`

mvadari commented on xrpl/models/transactions/transaction.py:
```diff
@@ -264,6 +267,10 @@ def _get_errors(self: Self) -> Dict[str, str]:
             account_txn_id must be None and sequence must be None or 0"""
 
+        if self.account == self.delegate:
+            error_msg = "Account and delegate addresses cannot be the same"
+            errors["Transaction"] = errors.get("Transaction", "") + error_msg
```
```suggestion
            errors["delegate"] = "Account and delegate addresses cannot be the same"
```

MustardseedX commented:
I am sorry i dident know i needed a computer. It was an honest and very
stupid mistake but in no way was it intentional

MustardseedXInK

On Thu, May 8, 2025, 09:17 coderabbitai[bot] ***@***.***>
wrote:

> *coderabbitai[bot]* left a comment (XRPLF/xrpl-py#840)
> <https://github.com/XRPLF/xrpl-py/pull/840#issuecomment-2863610813>
>
> Important
> Review skipped
>
> Draft detected.
>
> Please check the settings in the CodeRabbit UI or the .coderabbit.yaml
> file in this repository. To trigger a single review, invoke the @coderabbitai
> review command.
>
> You can disable this status message by setting the reviews.review_status
> to false in the CodeRabbit configuration file.
> ------------------------------
>
> Thanks for using CodeRabbit! It's free for OSS, and your support helps us
> grow. If you like it, consider giving us a shout-out.
> ❤️ Share
>
>    - X
>    <https://twitter.com/intent/tweet?text=I%20just%20used%20%40coderabbitai%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20the%20proprietary%20code.%20Check%20it%20out%3A&url=https%3A//coderabbit.ai>
>    - Mastodon
>    <https://mastodon.social/share?text=I%20just%20used%20%40coderabbitai%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20the%20proprietary%20code.%20Check%20it%20out%3A%20https%3A%2F%2Fcoderabbit.ai>
>    - Reddit
>    <https://www.reddit.com/submit?title=Great%20tool%20for%20code%20review%20-%20CodeRabbit&text=I%20just%20used%20CodeRabbit%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20proprietary%20code.%20Check%20it%20out%3A%20https%3A//coderabbit.ai>
>    - LinkedIn
>    <https://www.linkedin.com/sharing/share-offsite/?url=https%3A%2F%2Fcoderabbit.ai&mini=true&title=Great%20tool%20for%20code%20review%20-%20CodeRabbit&summary=I%20just%20used%20CodeRabbit%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20proprietary%20code>
>
> 🪧 Tips Chat
>
> There are 3 ways to chat with CodeRabbit
> <https://coderabbit.ai?utm_source=oss&utm_medium=github&utm_campaign=XRPLF/xrpl-py&utm_content=840>
> :
>
>    - Review comments: Directly reply to a review comment made by
>    CodeRabbit. Example:
>       - I pushed a fix in commit <commit_id>, please review it.
>       - Generate unit testing code for this file.
>       - Open a follow-up GitHub issue for this discussion.
>    - Files and specific lines of code (under the "Files changed" tab):
>    Tag @coderabbitai in a new review comment at the desired location with
>    your query. Examples:
>       - @coderabbitai generate unit testing code for this file.
>       - @coderabbitai modularize this function.
>    - PR comments: Tag @coderabbitai in a new PR comment to ask questions
>    about the PR branch. For the best results, please provide a very specific
>    query, as very limited context is provided in this mode. Examples:
>       - @coderabbitai gather interesting stats about this repository and
>       render them as a table. Additionally, render a pie chart showing the
>       language distribution in the codebase.
>       - @coderabbitai read src/utils.ts and generate unit testing code.
>       - @coderabbitai read the files in the src/scheduler package and
>       generate a class diagram using mermaid and a README in the markdown format.
>       - @coderabbitai help me debug CodeRabbit configuration file.
>
> Support
>
> Need help? Create a ticket on our support page
> <https://www.coderabbit.ai/contact-us/support> for assistance with any
> issues or questions.
>
> Note: Be mindful of the bot's finite context window. It's strongly
> recommended to break down tasks such as reading entire modules into smaller
> chunks. For a focused discussion, use review comments to chat about
> specific files and their changes, instead of using the PR comments.
> CodeRabbit Commands (Invoked using PR comments)
>
>    - @coderabbitai pause to pause the reviews on a PR.
>    - @coderabbitai resume to resume the paused reviews.
>    - @coderabbitai review to trigger an incremental review. This is
>    useful when automatic reviews are disabled for the repository.
>    - @coderabbitai full review to do a full review from scratch and
>    review all the files again.
>    - @coderabbitai summary to regenerate the summary of the PR.
>    - @coderabbitai generate docstrings to generate docstrings
>    <https://docs.coderabbit.ai/finishing-touches/docstrings> for this PR.
>    - @coderabbitai generate sequence diagram to generate a sequence
>    diagram of the changes in this PR.
>    - @coderabbitai resolve resolve all the CodeRabbit review comments.
>    - @coderabbitai configuration to show the current CodeRabbit
>    configuration for the repository.
>    - @coderabbitai help to get help.
>
> Other keywords and placeholders
>
>    - Add @coderabbitai ignore anywhere in the PR description to prevent
>    this PR from being reviewed.
>    - Add @coderabbitai summary to generate the high-level summary at a
>    specific location in the PR description.
>    - Add @coderabbitai anywhere in the PR title to generate the title
>    automatically.
>
> Documentation and Community
>
>    - Visit our Documentation <https://docs.coderabbit.ai> for detailed
>    information on how to use CodeRabbit.
>    - Join our Discord Community <http://discord.gg/coderabbit> to get
>    help, request features, and share feedback.
>    - Follow us on X/Twitter <https://twitter.com/coderabbitai> for
>    updates and announcements.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/XRPLF/xrpl-py/pull/840#issuecomment-2863610813>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3UTM2N3NA6VFWIK6WVWIUT25N7SDAVCNFSM6AAAAAB4W5M2BWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNRTGYYTAOBRGM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

mvadari commented on tests/integration/transactions/test_delegate_set.py:
```diff
@@ -0,0 +1,178 @@
+        ledger_index = "validated"
+        marker = None
+        granted_permission = set()
+        while True:
```
Same as in JS - why is this necessary? Much more efficient to search via `account_objects`.

Patel-Raj11 replied:
@mvadari I wanted to test for ledger_data api call. So I added this. But, I should also test for `account_objects` as your review comment pointed out missing DELEGATE enum type in `AccountObjectType`

mvadari replied:
I'm not sure this test is worth it - it'll take a while and be pretty inefficient.

Patel-Raj11 replied:
I see, removed `ledger_data` and added `account_objects` test.

mvadari commented on tests/unit/core/binarycodec/fixtures/data/serialized-dict-fixtures.json:
```diff
@@ -0,0 +1,49 @@
+            "TransactionType": "OfferCreate",
+            "TxnSignature": "304502202ABE08D5E78D1E74A4C18F2714F64E87B8BD57444AFA5733109EB3C077077520022100DB335EE97386E4C0591CAC024D50E9230D8F171EEB901B5E5E4BD6D1E0AEF98C"           
+        },
+        "buffer": "120007220000000024000195F964400000170A53AC2065D5460561EC9DE000000000000000000000000000494C53000000000092D705968936C419CE614BF264B5EEB1CEA47FF468400000000000000A7321028472865AF4CB32AA285834B57576B7290AA8C31B459047DB27E16F418D6A71667447304502202ABE08D5E78D1E74A4C18F2714F64E87B8BD57444AFA5733109EB3C077077520022100DB335EE97386E4C0591CAC024D50E9230D8F171EEB901B5E5E4BD6D1E0AEF98C8114 ...
```
To match the other fixture files, this should be called `binary`.

mvadari commented on tests/unit/core/binarycodec/fixtures/data/serialized-dict-fixtures.json:
```diff
@@ -0,0 +1,49 @@
+        },
+        "buffer": "120040210000F7E0228000000024000009186840000000000000C87321ED510865F867CDFCB944D435812ACF23D231E5C14534B146BCE46A2F794D198B777440D05A89D0B489DEC1CECBE0D33BA656C929CDCCC75D4D41B282B378544975B87A70C3E42147D980D1F6E2E4DC6316C99D7E2D4F6335F147C71C0DAA0D6516150D8114DB9157872FA63FAF7432CD300911A43B981B85A28514EBA79C385B47C50D52445DF2676EEC0231F732CEF01DEF203400000001E1EF203400000015E1 ...
+    }
+]
```
nit: newline at the end of the file

mvadari commented on tests/unit/core/binarycodec/fixtures/data/serialized-dict-fixtures.json:
```diff
@@ -0,0 +1,49 @@
+    },
+    {
+        "description": "DelegateSet Transaction",
+        "json": {
```
This should be added to the other fixtures file, not this one - `codec-fixtures.json`

mvadari commented on tests/unit/models/transactions/test_payment.py:
```diff
@@ -189,3 +190,13 @@ def test_mpt_payment(self):
         tx = Payment(**transaction_dict)
         self.assertTrue(tx.is_valid())
+
+    def test_payment_with_delegate_account(self):
```
Shouldn't this test be in `test_transaction.py`?

mvadari commented on xrpl/core/binarycodec/definitions/definitions.py:
```diff
@@ -59,6 +59,29 @@ def load_definitions(filename: str = "definitions.json") -> Dict[str, Any]:
+    "MPTokenIssuanceLock": 65547,
+    "MPTokenIssuanceUnlock": 65548,
+}
+_DELEGATABLE_PERMISSIONS_STR_TO_CODE_MAP: Dict[str, int] = {
+    **{key: value + 1 for (key, value) in _DEFINITIONS["TRANSACTION_TYPES"].items()},
+    **_GRANULAR_PERMISSIONS,
+}
+_DELEGATABLE_PERMISSIONS_CODE_TO_STR_MAP: Dict[int, str] = {
+    **{value + 1: key for (key, value) in _DEFINITIONS["TRANSACTION_TYPES"].items()},
+    **{value: key for (key, value) in _GRANULAR_PERMISSIONS.items()},
+}
```
nit (I think this is slightly better but it's fine if you don't want to change this):
```suggestion
_tx_delegations = {key: value + 1 for (key, value) in _DEFINITIONS["TRANSACTION_TYPES"].items()}
_DELEGATABLE_PERMISSIONS_STR_TO_CODE_MAP: Dict[str, int] = {
    **_tx_delegations,
    **_GRANULAR_PERMISSIONS,
}
_DELEGATABLE_PERMISSIONS_CODE_TO_STR_MAP: Dict[int, str] = {
    **{value: key for (key, value) in _DELEGATABLE_PERMISSIONS_STR_TO_CODE_MAP.items()}
}
```

Patel-Raj11 commented:
@mvadari I have addressed all the comments in [13f3eaa](https://github.com/XRPLF/xrpl-py/pull/840/commits/13f3eaaa5a24727e59e0363665b2f389bee58c84), can you please have a look?

mvadari commented:
Nice job
